Performance improvements for various Sets and Maps, and fix ChampHashMap's treatment of value overwriting and null values

retronym opened the pull request:
- Fast paths for `size/isEmpty`
 - Special case `isEmpty` when creating iterators
 - Avoid creating temporary objects to track the `{unmodified, modified, replaced}` states of Champ edits
 - Allow null values in ChampHashMap
 - When replacing an existing binding with a value that is `==` but not `eq`, we do need to perform an update.

retronym commented:
@msteindorfer could you please review the changes to `ChampHash{Map,Set}`?

retronym committed:
Override isEmpty in immutable.Set to avoid iterator creation

retronym committed:
Optimize TreeSet/Map isEmpty

retronym committed:
Avoid allocating {Map,Set}Effect CHAMP in update/remove

These were threaded through the update/remove process to track
whether a modification was made, and if so, whether that modification
was an update to an existing map binding.

In algorithms that interact with Maps in tight loops, especially
performing no-op add/removes, these temporary object can create
heap pressure. I noticed this while tuning the DPLL implementation
in the pattern matcher.

The first concern can be handled more efficiently by checking
reference equality.

The second is rather hard to deal with without VM support for
multiple return / multi-param value classes.

The options seemed to range between:

  - Pass in a mutable result object (the status quo)
  - Encode the "replaced-ness" in the MapNode itself (which is hard
    to do without increasing its footprint or without introducing more
    subclasses and compromising bi-morphic JIT inlining)
  - redundantly re-querying the key in the input node after updated
    returns a modified node, which requires extra calls to key.hashCode
    and extra traversal through the data structure.

This commit tries a different approach. We loosen the return type
of `updated` from `MapNode` to `MapNodeSource` (of which `MapNode` is
a subclass). When an update actually does overwrite an existing
binding, we return a different subclass, `Replaced`, which wraps the
actual node and represent the extra bit of information.

This could be seen as a an encoding of `Either[MapNode, MapNode]`,
which is optimized for zero allocations in the hot paths.

In the process, I also noticed and fixed a problem with maps
failing to replace a binding if the new value was equal but
not the same reference as the previous one.

Encode replaced-ness with by returning a boxed MapNode

retronym committed:
Add a fast path for empty Champ collections

retronym committed:
Efficient implemnentation of keys/valuesInterator in ChampHashMap

retronym committed:
Optimize mutable.HashSet

retronym committed:
Avoid more temporary Tuples in ChampMap

msteindorfer commented:
@retronym, yes, I'll have a look and add a review.

msteindorfer commented on src/library/scala/collection/immutable/ChampHashMap.scala:
```diff
@@ -170,8 +194,8 @@ private final class BitmapIndexedMapNode[K, +V](val dataMap: Int, val nodeMap: I
       val index = indexFrom(dataMap, mask, bitpos)
-      val payload = this.getPayload(index)
-      return if (key == payload._1) Option(payload._2) else Option.empty
+      val payload = this.getKey(index)
```
Since the semantic of the variable changed, you could consider renaming `payload` to e.g. `key0`.

msteindorfer replied:
You might still consider renaming the variable name before merging.

msteindorfer commented on src/library/scala/collection/immutable/ChampHashMap.scala:
```diff
@@ -205,7 +229,8 @@ private final class BitmapIndexedMapNode[K, +V](val dataMap: Int, val nodeMap: I
       val index = indexFrom(dataMap, mask, bitpos)
-      return (key, value) == this.getPayload(index)
+      // TODO test drive the change from == to eq here.
+      key == this.getKey(index) && (value.asInstanceOf[AnyRef] eq this.getValue(index).asInstanceOf[AnyRef])
```
What's the reasoning for using `eq`?

retronym replied:
The only time we should `Object.equals` (aka `AnyRef.==`) of the values is from `Map.hashCode,equals`.

 - to be conservative about not knowing whether it is a cheap or expensive operation, and
 - because the object identity is an observable, we the property that `map.update(k, v).apply(k) eq v` should hold, as the user might depend on object identity (e.g. by storing some state in fields that don't factor into equality)

msteindorfer replied:
`map.update(k, v).apply(k) eq v` holds anyways, since `update` always overrides the old `v` and since `apply` calls `containsKey` (and not `contains` which you edited above, see [link](https://github.com/retronym/scala/blob/71ac5272960bdd4fb8fa5521465865e6c892ca6e/src/library/scala/collection/immutable/ChampHashMap.scala#L57)). The purpose of `contains` is to check if key/value pair is present (could be also named `containsTuple`).

Maybe I'm missing the obvious, but I still don't see the need and correctness for this change. The following snippet would fail with the change above:

```
case class A(a: Int)

assert Map().updated(1, A(1)).contains(1, A(1)) // fails since both `A(1) ne A(1)`

```

retronym replied:
At least in the baseline revision, there is a bug for collision nodes:

```
scala> case class C(equalityElem: Int)(val other: Int)
defined class C

scala> assert(C(0)(0) == C(0)(1))
```

```
scala> class Key(i: Int) { override val hashCode: Int = 0; override val toString = s"Collide($i)" }
defined class Key

scala> val map = collection.immutable.ChampHashMap[Key, C]().updated(new Key(0), null).updated(new Key(1), null)
map: scala.collection.immutable.ChampHashMap[Key,C] = ChampHashMap(Collide(0) -> null, Collide(1) -> null)

scala> val k2 = new Key(2); map.updated(k2, C(0)(0)).updated(k2, C(0)(1)).apply(k2).other
k2: Key = Collide(2)
res9: Int = 0
```

That comes down to: https://github.com/scala/scala/blob/43bf2351ec0f251f4d6cdc5a0fdac75c7ea6ce1e/src/library/scala/collection/immutable/ChampHashMap.scala#L503-L506


For regular nodes, the overwritten value does appear to win:

```
scala> val map = collection.immutable.ChampHashMap[Key, C]()
map: scala.collection.immutable.ChampHashMap[Key,C] = ChampHashMap()

scala> val k2 = new Key(2); map.updated(k2, C(0)(0)).updated(k2, C(0)(1)).apply(k2).other
k2: Key = Collide(2)
res10: Int = 1
```

With this PR:

```
scala> val map = collection.immutable.ChampHashMap[Key, C]()
map: scala.collection.immutable.ChampHashMap[Key,C] = ChampHashMap()

scala> val k2 = new Key(2); map.updated(k2, C(0)(0)).updated(k2, C(0)(1)).apply(k2).other
k2: Key = Collide(2)
res0: Int = 1

scala> val map = collection.immutable.ChampHashMap[Key, C]().updated(new Key(0), null).updated(new Key(1), null)
map: scala.collection.immutable.ChampHashMap[Key,C] = ChampHashMap(Collide(0) -> null, Collide(1) -> null)

scala> val k2 = new Key(2); map.updated(k2, C(0)(0)).updated(k2, C(0)(1)).apply(k2).other
k2: Key = Collide(2)
res1: Int = 1
```

>  The following snippet would fail with the change above:

That code doesn't typecheck against the `ChampHashMap` API, as only the internal `MapNode` has the `contains(key, value, ..)` API. 

What seems to be missing, before and after this PR, is sharing of the map when the new value really `eq` the exisitng one.

```
scala> case class A(a: Int); val map = collection.immutable.ChampHashMap[Int, A](); val a = A(1); val map1 = map.updated(1, a); val map2 = map1.updated(1, a); map1 eq map2
defined class A
map: scala.collection.immutable.ChampHashMap[Int,A] = ChampHashMap()
a: A = A(1)
map1: scala.collection.immutable.ChampHashMap[Int,A] = ChampHashMap(1 -> A(1))
map2: scala.collection.immutable.ChampHashMap[Int,A] = ChampHashMap(1 -> A(1))
res0: Boolean = false
```

msteindorfer commented on src/library/scala/collection/immutable/ChampHashMap.scala:
```diff
@@ -563,17 +590,41 @@ private final class HashCollisionMapNode[K, +V](val hash: Int, val content: Vect
 
 private final case class MapEffect[K, +V]() {
```
I assume that `MapEffect` is obsolete your revision and should then be removed for consistency?

msteindorfer reviewed (changes requested):
@retronym, I reviewed the `immutable.Hash{Set, Map}` part of this PR (I didn't look at TreeMap, etc).

In general, I see where you're heading towards. 

* The addition of specialized `getKey`, `getValue`, `MapKeyIterator`, `MapValueIterator` are fine. The follow exactly how I designed the implementation to be extended. I have the same specializations in Capsule, but left them out to keep the code base of the collections concise.

* In my experience, it's better for performance / cache locality to keep using `getPayload` when you eventually will need both key and value. I'm not sure about Scala, but check / verify if the JVM's escape analysis eventually gets rid off the allocation of the tuple. Otherwise, explicitly using `getKey` and `getValue` separate as you do is fine as well.

* For `ChampHashSet` the `SetEffect` abstraction isn't necessary and can safely be replaced by `ne` / `eq` for comparing root nodes. I kept it mostly for consistency between set and map implementation. See also caveat mentioned in next bullet point.

* There exist optimisations for persistent data structures that utilize internal mutability for improving the performance of batch operations (e.g., insert all, map, builders, ...). In such scenarios the identity of the root node may remain unchanged, even under modifications. This is related to Clojure's `transients` for example and doesn't apply to the current `ChampHash{Set, Map}` implementations in Scala, but should be kept in mind when merging this PR.

* For `ChampsHashSet` the `MapEffect` captures three different states that cannot be modelled by `ne` / `eq` comparisons on the root node. I would advise (strongly) against introducing new sub-classes (i.e., `*NodeReplaced`) for modelling the third state, since this would rather hurt performance overall and IMO is not clean either. The JVM can optimize method dispatch well up to bimorphic call sites (BitmapIndexedNode and HashCollisionNode). With more types, method dispatch will be always megamorphic and not optimised at all by the JVM. Further, conceptually replacing a binding is an (temporary) effect of the update operation and not a (permanent) property of the node.

I'm in doubt which way to _optimize_ to remove `MapEffect`. The old `HashMap` didn't had this issue, since it cached the size in every `MapNode`, but on the other hand had much higher memory footprints therefore.

msteindorfer commented:
@retronym what was the overall motivation of implementing the refactoring in this PR? Was there a need (e.g., benchmark results, complaints, etc)?

retronym commented:
I was optimizing part of the compiler in https://github.com/scala/scala/pull/6598. I was compiling a source file that stresses our pattern matching exhaustivity analysis, which uses immutable maps internally. I noticed that when using `ChampHashMap` in the tight loop, Java Flight Recorder reported that `MapEffect` was the dominant allocation (so escape analysis in HotSpot C2 didn't seem capable to eliminate it).

I then reviewed the implementation to understand the role of these, and thought it looked possible to remove them.

You make a good point about being mindful of bimorphic inlining. I'll run some benchmarks with JITWatch to try to see this problem in action. I would have liked to just spend a field in MapNode for this info, but unfortunately adding a field to:

```scala
class BitmapIndexedMapNode[K, +V](val dataMap: Int, val nodeMap: Int, val content: Array[Any])
```

Will bump it up to a bigger object size.

Are there any bits free in `dataMap` or `nodeMap`?

I do share your reservations about having the map remember the "replaced?" fact from its creation. But I have bigger reservations about allocating those temporary objects each call to `excl`/`incl`. They will never make it out of the young generation, but when called in a tight loop they can drastically increase the rate of young generation collections, which can hasten promotion of other objects that have small, but not small enough, lifetimes.

msteindorfer commented:
Thanks for the explanation and the context.

> Are there any bits free in dataMap or nodeMap?

No free bits unfortunately.

You could also consider calling `rootNode.contains(key, keyHash, 0)` to get the additional information you require. It's another trade-off since it will slow down updates (at least in some branches), but would allow removing `MapResult` without the `*Replaced` classes / methods. You might give it a try in compiler benchmarks, maybe it's not that bad.

It's very unsatisfying dealing with this design issue, I have to agree.

julienrf commented:
@retronym Another thing I had in mind was to override operations that return collections (`map`, `filter`, etc.) to use a builder based approach instead of a view based approach, for strict collection types. We did that in `Seq` and got a [performance improvement of ~25% over view based implementations](https://www.scala-lang.org/blog/2018/02/09/collections-performance.html). (see also https://github.com/scala/collection-strawman/issues/325#issuecomment-360412201 which mentions this idea)

retronym commented:
@msteindorfer I've pushed commits implementing the `Replaced` idea.

msteindorfer commented:
@retronym I'll try to have a detailed look in the evening, but the `Replaced` wrapper seems to be a good alternative solution to minimise allocations for communicating the effect.

msteindorfer reviewed (changes requested):
Looks good in general, see code comments for minor discussion points.

retronym committed:
Reuse existing hashmap when overwriting a binding with an identical value

retronym committed:
Remove dead code in HashMap

retronym committed:
Fix handling of null values in ChampHashMap

retronym commented on src/library/scala/collection/immutable/ChampHashMap.scala:
```diff
@@ -170,16 +197,16 @@ private final class BitmapIndexedMapNode[K, +V](val dataMap: Int, val nodeMap: I
     if ((dataMap & bitpos) != 0) {
       val index = indexFrom(dataMap, mask, bitpos)
-      val payload = this.getPayload(index)
-      return if (key == payload._1) Option(payload._2) else Option.empty
```
@msteindorfer Not sure if you're aware, but `Option(null)` == `None`. That's not what's wanted here, as we can see from:

```
Welcome to Scala 2.13.0-M4 (Java HotSpot(TM) 64-Bit Server VM, Java 1.8.0_172).
Type in expressions for evaluation. Or try :help.

scala> val map = collection.immutable.ChampHashMap[Any, Any]()
map: scala.collection.immutable.ChampHashMap[Any,Any] = ChampHashMap()

scala> map.updated("", null).get("")
res0: Option[Any] = None

scala> val map = collection.immutable.HashMap[Any, Any]()
map: scala.collection.immutable.HashMap[Any,Any] = HashMap()

scala> map.updated("", null).get("")
res1: Option[Any] = Some(null)
```

msteindorfer replied:
My mistake. I used `Option.empty` wrongly, thought it construct `None`.

retronym replied:
It's `Option.apply` which is something of a trap. It has the same semantics Java's `Optional.ofNullable`.

```
scala> Option.empty == None
res1: Boolean = true

scala> Option.apply("") == Some("")
res2: Boolean = true

scala> Option.apply(null) == None
res3: Boolean = true
```

msteindorfer reviewed (approved):
@retronym, I think we discussed all issues about communicating back the states and also about identity and map-value replacement.

Please go ahead and incorporate the suggested changes, and maybe tweak the map behaviour on `eq` on values.

mkeskells commented:
@retronym there is at least one place where the jit will not elide the zeroing on the array after creation  -  https://github.com/retronym/scala/blob/c531621fe8f324de135d967c8ea673759e16469e/src/library/scala/collection/immutable/ChampHashMap.scala#L447, as this seperates the create from the copy

I have not confirmed all of the other cases generate the required bytecode pattern. Not sure what the magic pattern is currently

for https://github.com/retronym/scala/blob/c531621fe8f324de135d967c8ea673759e16469e/src/library/scala/collection/immutable/ChampHashMap.scala#L369 this seems closer to Arrays.copyOf

retronym committed:
Move array creation next to arraycopy to help JIT

JIT can elide array zeroing it can prove that ranges of the
array will be written to before the reference to the array
escapes. We should avoiding code between the instantation
and the arraycopy(s) to make the pattern of code clear to JIT.

viktorklang commented on src/library/scala/collection/immutable/ChampHashSet.scala:
```diff
@@ -32,28 +32,29 @@ final class ChampHashSet[A] private[immutable] (val rootNode: SetNode[A], val ca
   override def isEmpty: Boolean = cachedSize == 0
 
-  def iterator: Iterator[A] = new SetIterator[A](rootNode)
+  def iterator: Iterator[A] = {
```
Nice

viktorklang commented on src/library/scala/collection/immutable/TreeSet.scala:
```diff
@@ -40,6 +40,8 @@ final class TreeSet[A] private (tree: RB.Tree[A, Unit])(implicit val ordering: O
 
   override def size: Int = RB.count(tree)
 
+  override def isEmpty = size == 0
```
`RB.isEmpty(tree)` is probably cheaper

Also, we'll most likely want to fixe the iterators in RedBlackTree since they allocate new iterators unconditionally: https://github.com/retronym/scala/blob/ec5d494534415fe8d4f2ecdec3bdb22c3a0d9487/src/library/scala/collection/immutable/RedBlackTree.scala#L138

viktorklang replied:
@retronym Not sure you saw these comments.

viktorklang commented on src/library/scala/collection/mutable/HashSet.scala:
```diff
@@ -25,7 +25,6 @@ final class HashSet[A]
   // Used by scala-java8-compat (private[mutable] erases to public, so Java code can access it)
   private[mutable] def getTable: FlatHashTable[A] = table
-
   override def iterator: Iterator[A] = table.iterator
```
This seems to unconditionally allocate iterator also for empty: https://github.com/retronym/scala/blob/ec5d494534415fe8d4f2ecdec3bdb22c3a0d9487/src/library/scala/collection/mutable/FlatHashTable.scala#L197

retronym commented:
Thanks for the comments, I’ll follow up on Monday. I expect there are still
plenty of wins like this around.
